File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    bool checkEqual(int sc[123], int tc[123]){
        for(int i=0; i<123; i++){
            if(sc[i] < tc[i]) return false;
        }
        return true;
    }
    string minWindow(string s, string t) {
        int sc[123] = {0}, tc[123] = {0};
        for(auto i: t) {
            if(isupper(i)) tc[i]++;
            else tc[i]++;
        }
        int i=0;
        while(i < s.length() && i < t.length()){
            if(isupper(s[i])) sc[s[i]]++;
            else sc[s[i]]++;
            i++;
        }
        if(checkEqual(sc,tc)) return s.substr(0, t.length());

        int f = 0;
        int mini = INT_MAX;
        int start = -1, end = -1;
        while(i < s.length()){
            if(isupper(s[i])) sc[s[i]]++;
            else sc[s[i]]++;
            
            while(checkEqual(sc,tc)){
                if((i-f+1) < mini){
                    mini = i-f+1;
                    start = f;
                    end = i;
                }
                if(isupper(s[f])) sc[s[f]]--;
                else sc[s[f]]--;
                f++;
            }
            i++;
        }
        if(start == -1) return "";
        return s.substr(start, end-start+1);
    }
};

File: 76-minimum-window-substring/minimum-window-substring.py
class Solution:
    def compare_lists(self, s_list, t_list):
        for i in range(128):
            if t_list[i] > 0 and s_list[i] < t_list[i]:
                return False
        return True


    def minWindow(self, s: str, t: str) -> str:

        if len(s) < len(t):
            return ""
        
        list1 = [0] * 128
        list2 = [0] * 128

        for ch in t:
            idx = ord(ch) - ord('A')
            list2[idx] += 1
        
        l = 0
        r = 0

        mini = 1e10
        start = end = -1

        while(r < len(s)):
            
            idx = ord(s[r]) - ord('A')
            list1[idx] += 1
            
            while (l <= r and self.compare_lists(list1, list2)):
                if (r-l+1) < mini:
                    mini = r-l+1
                    start = l
                    end = r
                idx = ord(s[l]) - ord('A')
                list1[idx] -= 1
                l += 1
            
            r += 1
        
        while (l < r and self.compare_lists(list1, list2)):
            if (r-l+1) < mini:
                mini = r-l+1
                start = l
                end = r
            idx = ord(s[l]) - ord('A')
            list1[idx] -= 1
            l += 1
        
        if start == -1:
            return ""
        return s[start:end+1]
        